// appcenter-cli/dist/commands/codepush/lib/environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Default environment data
const codePushEnvironmentsData = {
    defaultEnvironment: "prod",
    environments: {
        dev: {
            acquisitionEndpoint: "https://codepush-int-legacy.azurewebsites.net/",
            managementEndpoint: "https://codepush-management-int-legacy.azurewebsites.net/",
            description: "Development"
        },
        int: {
            acquisitionEndpoint: "https://codepush-int-legacy.azurewebsites.net/",
            managementEndpoint: "https://codepush-management-int-legacy.azurewebsites.net/",
            description: "Integration"
        },
        staging: {
            acquisitionEndpoint: "https://codepush-staging.azurewebsites.net/",
            managementEndpoint: "https://codepush-management-staging.azurewebsites.net/",
            description: "Staging"
        },
        prod: {
            acquisitionEndpoint: "https://codepush.azurewebsites.net/",
            managementEndpoint: "https://codepush-management.azurewebsites.net/",
            description: "Production"
        }
    }
};
function environments(environmentName = codePushEnvironmentsData.defaultEnvironment) {
    return codePushEnvironmentsData.environments[environmentName];
}
exports.environments = environments;
