// appcenter-cli/dist/commands/distribute/groups/update.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const commandline_1 = require("../../../util/commandline");
const apis_1 = require("../../../util/apis");
const interaction_1 = require("../../../util/interaction");
const util_1 = require("util");
const _ = require("lodash");
const list_of_users_helper_1 = require("../../../util/misc/list-of-users-helper");
const debug = require("debug")("appcenter-cli:commands:distribute:groups:update");
let UpdateDistributionGroupCommand = class UpdateDistributionGroupCommand extends commandline_1.AppCommand {
    run(client) {
        return __awaiter(this, void 0, void 0, function* () {
            const app = this.app;
            // validate that string and file properties are not specified simultaneously
            this.validateParameters();
            // validating parameters and loading provided files (if any)
            const testersToAdd = list_of_users_helper_1.getUsersList(this.testersToAdd, this.testersToAddListFile, debug);
            const testersToDelete = list_of_users_helper_1.getUsersList(this.testersToDelete, this.testersToDeleteListFile, debug);
            const newDistributionGroupNameValidation = this.isDistributionGroupNameFree(client, app, this.newDistributionGroupName);
            // showing spinner while parameters are validated
            const [testersToAddEmails, testersToDeleteEmails] = yield interaction_1.out.progress("Validating parameters...", Promise.all([testersToAdd, testersToDelete, newDistributionGroupNameValidation]));
            let deletedTestersEmails;
            if (testersToDeleteEmails.length) {
                debug("Deleting testers from distribution group");
                deletedTestersEmails = yield this.deleteTestersFromDistributionGroup(client, app, testersToDeleteEmails);
            }
            else {
                deletedTestersEmails = [];
            }
            let addedTestersEmails;
            if (testersToAddEmails.length) {
                debug("Adding testers to distribution group");
                addedTestersEmails = yield this.addTestersToDistributionGroup(client, app, testersToAddEmails);
            }
            else {
                addedTestersEmails = [];
            }
            let currentGroupName;
            if (!_.isNil(this.newDistributionGroupName)) {
                debug("Renaming the distribution group");
                currentGroupName = yield this.renameDistributionGroup(client, app, this.newDistributionGroupName);
            }
            else {
                currentGroupName = this.distributionGroup;
            }
            if (deletedTestersEmails.length !== testersToDeleteEmails.length || addedTestersEmails.length !== testersToAddEmails.length) {
                interaction_1.out.text("Updating the list of testers was partially successful");
            }
            interaction_1.out.text((result) => `Distribution group ${result.name} was successfully updated`, { name: currentGroupName, addedTesters: addedTestersEmails, deletedTesters: deletedTestersEmails });
            return commandline_1.success();
        });
    }
    validateParameters() {
        if (_.isNil(this.newDistributionGroupName) && _.isNil(this.testersToAdd) && _.isNil(this.testersToAddListFile) && _.isNil(this.testersToDelete) && _.isNil(this.testersToDeleteListFile)) {
            throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, "nothing to update");
        }
        if (!_.isNil(this.testersToAdd) && !_.isNil(this.testersToAddListFile)) {
            throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, "parameters 'add-testers' and 'add-testers-file' are mutually exclusive");
        }
        if (!_.isNil(this.testersToDelete) && !_.isNil(this.testersToDeleteListFile)) {
            throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, "parameters 'delete-testers' and 'delete-testers-file' are mutually exclusive");
        }
    }
    isDistributionGroupNameFree(client, app, name) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!_.isNil(name)) {
                try {
                    const httpRequest = yield apis_1.clientRequest((cb) => client.distributionGroups.get(app.ownerName, app.appName, name, cb));
                    // Throw an exception if 404 error was not thrown during clientRequest
                    throw httpRequest.response.statusCode;
                }
                catch (error) {
                    if (error && error.response && error.response.statusCode === 404) {
                        // 404 is correct status code for this case
                        return;
                    }
                    if (error === 200) {
                        throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, `distribution group ${name} already exists`);
                    }
                    else {
                        debug(`Failed to check if the distribution group ${name} already exists - ${util_1.inspect(error)}`);
                        throw commandline_1.failure(commandline_1.ErrorCodes.Exception, `failed to check if the distribution group ${name} already exists`);
                    }
                }
            }
        });
    }
    deleteTestersFromDistributionGroup(client, app, userEmails) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const httpResponse = yield interaction_1.out.progress("Deleting testers from the distribution group...", apis_1.clientRequest((cb) => client.distributionGroups.removeUser(app.ownerName, app.appName, this.distributionGroup, {
                    userEmails
                }, cb)));
                if (httpResponse.response.statusCode >= 400) {
                    throw httpResponse.response.statusCode;
                }
                else {
                    return httpResponse.result.filter((result) => result.status < 400).map((result) => result.userEmail);
                }
            }
            catch (error) {
                if (error === 404) {
                    throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, `distribution group ${this.distributionGroup} doesn't exist`);
                }
                else {
                    debug(`Failed to delete testers from the distribution group ${this.distributionGroup} - ${util_1.inspect(error)}`);
                    throw commandline_1.failure(commandline_1.ErrorCodes.Exception, `failed to delete testers from the distribution group`);
                }
            }
        });
    }
    addTestersToDistributionGroup(client, app, userEmails) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const httpResponse = yield interaction_1.out.progress("Adding testers to the distribution group...", apis_1.clientRequest((cb) => client.distributionGroups.addUser(app.ownerName, app.appName, this.distributionGroup, {
                    userEmails
                }, cb)));
                if (httpResponse.response.statusCode >= 400) {
                    throw httpResponse.response.statusCode;
                }
                else {
                    return httpResponse.result.filter((result) => result.status < 400).map((result) => result.userEmail);
                }
            }
            catch (error) {
                if (error === 404) {
                    throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, `distribution group ${this.distributionGroup} doesn't exist`);
                }
                else {
                    debug(`Failed to add testers to the distribution group ${this.distributionGroup} - ${util_1.inspect(error)}`);
                    throw commandline_1.failure(commandline_1.ErrorCodes.Exception, "failed to add testers to the distribution group");
                }
            }
        });
    }
    renameDistributionGroup(client, app, newName) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const httpResponse = yield interaction_1.out.progress("Renaming the distribution group...", apis_1.clientRequest((cb) => client.distributionGroups.update(app.ownerName, app.appName, this.distributionGroup, {
                    name: newName,
                }, cb)));
                if (httpResponse.response.statusCode >= 400) {
                    throw httpResponse.response.statusCode;
                }
                else {
                    return newName;
                }
            }
            catch (error) {
                switch (error) {
                    case 400:
                        throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, `${this.distributionGroup} group can't be renamed`);
                    case 404:
                        throw commandline_1.failure(commandline_1.ErrorCodes.InvalidParameter, `distribution group ${this.distributionGroup} doesn't exist`);
                    default:
                        debug(`Failed to rename the distribution group ${this.distributionGroup} - ${util_1.inspect(error)}`);
                        throw commandline_1.failure(commandline_1.ErrorCodes.Exception, `failed to rename the distribution group`);
                }
            }
        });
    }
};
__decorate([
    commandline_1.help("Distribution group name"),
    commandline_1.shortName("g"),
    commandline_1.longName("group"),
    commandline_1.required,
    commandline_1.hasArg
], UpdateDistributionGroupCommand.prototype, "distributionGroup", void 0);
__decorate([
    commandline_1.help("New distribution group name"),
    commandline_1.shortName("n"),
    commandline_1.longName("name"),
    commandline_1.hasArg
], UpdateDistributionGroupCommand.prototype, "newDistributionGroupName", void 0);
__decorate([
    commandline_1.help("List of testers to add (use space-separated list of e-mails)"),
    commandline_1.shortName("t"),
    commandline_1.longName("add-testers"),
    commandline_1.hasArg
], UpdateDistributionGroupCommand.prototype, "testersToAdd", void 0);
__decorate([
    commandline_1.help("List of testers to delete (use space-separated list of e-mails)"),
    commandline_1.shortName("d"),
    commandline_1.longName("delete-testers"),
    commandline_1.hasArg
], UpdateDistributionGroupCommand.prototype, "testersToDelete", void 0);
__decorate([
    commandline_1.help("Path to file containing list of testers to add"),
    commandline_1.shortName("T"),
    commandline_1.longName("add-testers-file"),
    commandline_1.hasArg
], UpdateDistributionGroupCommand.prototype, "testersToAddListFile", void 0);
__decorate([
    commandline_1.help("Path to file containing list of testers to delete"),
    commandline_1.shortName("D"),
    commandline_1.longName("delete-testers-file"),
    commandline_1.hasArg
], UpdateDistributionGroupCommand.prototype, "testersToDeleteListFile", void 0);
UpdateDistributionGroupCommand = __decorate([
    commandline_1.help("Update existing distribution group")
], UpdateDistributionGroupCommand);
exports.default = UpdateDistributionGroupCommand;
