// appcenter-cli/dist/commands/crashes/lib/azure-blob-upload-helper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const AzureStorage = require("azure-storage");
const Url = require("url");
const commandline_1 = require("../../../util/commandline");
const util_1 = require("util");
class AzureBlobUploadHelper {
    constructor(debug) {
        this.debug = debug;
    }
    upload(uploadUrl, zip) {
        return __awaiter(this, void 0, void 0, function* () {
            const urlObject = Url.parse(uploadUrl);
            const blobService = this.getBlobService(urlObject);
            const [container, blob] = this.getContainerAndBlob(urlObject);
            yield this.uploadBlockBlob(blobService, container, blob, zip);
        });
    }
    uploadBlockBlob(blobService, container, blob, file) {
        return new Promise((resolve, reject) => {
            blobService.createBlockBlobFromLocalFile(container, blob, file, {
                contentSettings: {
                    contentType: "application/zip"
                }
            }, (error, result, response) => {
                if (error) {
                    this.debug(`Failed to upload ZIP with symbols - ${util_1.inspect(error)}`);
                    reject(commandline_1.failure(commandline_1.ErrorCodes.Exception, "failed to upload ZIP with symbols"));
                }
                else {
                    resolve();
                }
            });
        });
    }
    getBlobService(urlObject) {
        const blobEndpoint = Url.format({
            protocol: urlObject.protocol,
            host: urlObject.host
        });
        const sharedAccessSignature = urlObject.query;
        const connectionString = "BlobEndpoint=" + blobEndpoint + ";" + "SharedAccessSignature=" + sharedAccessSignature;
        return new AzureStorage.BlobService(connectionString);
    }
    getContainerAndBlob(urlObject) {
        const splitPathName = urlObject.pathname.split("/");
        return [splitPathName[1], splitPathName[2]];
    }
}
exports.default = AzureBlobUploadHelper;
