// appcenter-cli/dist/util/commandline/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var command_1 = require("./command");
exports.Command = command_1.Command;
var app_command_1 = require("./app-command");
exports.AppCommand = app_command_1.AppCommand;
exports.getCurrentApp = app_command_1.getCurrentApp;
var option_decorators_1 = require("./option-decorators");
exports.shortName = option_decorators_1.shortName;
exports.longName = option_decorators_1.longName;
exports.hasArg = option_decorators_1.hasArg;
exports.position = option_decorators_1.position;
exports.name = option_decorators_1.name;
exports.required = option_decorators_1.required;
exports.defaultValue = option_decorators_1.defaultValue;
exports.help = option_decorators_1.help;
var command_runner_1 = require("./command-runner");
exports.runner = command_runner_1.runner;
__export(require("./command-result"));
__export(require("./autocomplete"));
