// appcenter-cli/dist/commands/test/lib/generate-command.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const commandline_1 = require("../../../util/commandline");
const help_messages_1 = require("../lib/help-messages");
const pfs = require("../../../util/misc/promisfied-fs");
class GenerateCommand extends commandline_1.Command {
    constructor(args) {
        super(args);
    }
    // Override this if you need to validate options
    validateOptions() {
        return __awaiter(this, void 0, void 0, function* () {
            return;
        });
    }
    isIOS() {
        return (this.platform.toLowerCase() === "ios");
    }
    isAndroid() {
        return (this.platform.toLowerCase() === "android");
    }
    copyTemplates() {
        return __awaiter(this, void 0, void 0, function* () {
            const templatePath = this.isIOS() ? this.templatePathiOS : this.templatePathAndroid;
            yield pfs.cpDir(templatePath, this.outputPath);
        });
    }
    runNoClient() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!(this.isIOS() || this.isAndroid())) {
                throw new Error("Valid values of argument --platform are 'ios' and 'android'");
            }
            yield this.validateOptions();
            if (yield pfs.exists(this.outputPath)) {
                const files = yield pfs.readdir(this.outputPath);
                if (!(files.length === 0)) {
                    return commandline_1.failure(commandline_1.ErrorCodes.Exception, this.outputPath + " exists and is not empty");
                }
            }
            yield this.copyTemplates();
            yield this.processTemplate();
            return commandline_1.success();
        });
    }
}
__decorate([
    commandline_1.help(help_messages_1.Messages.TestCloud.Arguments.AppPlatform),
    commandline_1.longName("platform"),
    commandline_1.required,
    commandline_1.hasArg
], GenerateCommand.prototype, "platform", void 0);
__decorate([
    commandline_1.help(help_messages_1.Messages.TestCloud.Arguments.GenerateOutputPath),
    commandline_1.longName("output-path"),
    commandline_1.required,
    commandline_1.hasArg
], GenerateCommand.prototype, "outputPath", void 0);
exports.GenerateCommand = GenerateCommand;
